DiskBBQ Updates (#4037)
elastic/docs-content: deploy-manage/production-guidance/optimize-performance/approximate-knn-search.md

## Ensure data nodes have enough memory [_ensure_data_nodes_have_enough_memory]

{{es}} uses either the [HNSW](https://arxiv.org/abs/1603.09320) algorithm or the [DiskBBQ](https://www.elastic.co/search-labs/blog/diskbbq-elasticsearch-introduction) algorithm for approximate kNN search. 

HNSW is a graph-based algorithm which only works efficiently when most vector data is held in memory. You should ensure that data nodes have at least enough RAM to hold the vector data and index structures.

> DiskBBQ is a clustering algorithm which can scale effeciently on a fraction of the total memory. You can start with enough RAM to hold the vector data and index structures but it's likely you will be able to use signifigantly less than this and still maintain good performance. In testing we find this will be between 1-5% of the index structure size (centroids and quantized vector data) per unique query where unique queries access non-overlapping clusters.  

Author: reworded this a bit let me know what you think now

If utilizing HNSW, the graph must also be in memory, to estimate the required bytes use `num_vectors * 4 * HNSW.m`. The default value for `HNSW.m` is 16, so by default `num_vectors * 4 * 16`.

> If utilizing DiskBBQ, a fraction of the clusters and centroids will need to be in memory.  When doing this estimation it makes more sense to include both the index structure and the quantized vectors together as the structures are dependent. To estimate the total bytes we compute the cost of the centroids as `num_clusters * num_dimensions * 4 + num_clusters * (num_dimensions + 14)` plus the cost of the quantized vectors within the clusters as `num_vectors * ((num_dimensions/8 + 14 + 2) * 2)` where `num_clusters` is defined as `num_vectors / vectors_per_cluster` which by default will be `num_vectors / 384`

Author: made an attempt to make this easier to read; let me know what ya'll think.

> If utilizing DiskBBQ, a fraction of the clusters and centroids will need to be in memory.  When doing this estimation it makes more sense to include both the index structure and the quantized vectors together as the structures are dependent. To estimate the total bytes we compute the cost of the centroids as `num_clusters * num_dimensions * 2 * 4 + num_clusters * (num_dimensions + 14)` plus the cost of the quantized vectors within the clusters as `num_vectors * ((num_dimensions/8 + 14 + 2) * 2)` where `num_clusters` is defined as `num_vectors / vectors_per_cluster` which by default will be `num_vectors / 384`

Author: That 2 is definitely not right.  Removing that. 

Went back and double checked what bytes are there.  

```
posting list w/o vectors = num_clusters * (centroid_bytes + dotProd(centroid,centroid) + clusterSize + encoding)
posting list w/o vectors = num_clusters * ((dims * 4) + 4 + 4 + 1)
```

I dont think it's worth it to describe the additional bytes.  That would just make this even more complicated.

Suggestions for simplifying it are welcome.  It probably seems more complex because it includes both the "index structure" for the centroids and the vectors because of SOAR.  Maybe I can reference the bbq quantized calculation from above and somewhat simplify the centroids figures.  So something like this instead.  Thoughts?

```
the cost of the centroids as `num_clusters * (num_dimensions * 4  + (num_dimensions + 14))` 
```

```
plus the cost of the quantized vectors within the clusters as `bbq_quantizated_vectors * 2` 
```

The downside is this ignores the doc id (2 byte) cost per vector.  I almost feel like it makes it more complicated to mention this rather than not if we break it up for instance like this: 

```
plus the cost of the quantized vectors within the clusters as `bbq_quantizated_vectors * 2` + `num_vectors * 2 * 2` 
```  

I'll think about it some more.

Loading data into the filesystem cache eagerly on too many indices or too many files will make search *slower* if the filesystem cache is not large enough to hold all the data. Use with caution.
::::

The following file extensions are used for the approximate kNN search: Each extension is broken down by the quantization types.

* {applies_to}`stack: ga 9.3` `cenivf` for DiskBBQ to store centroids
* {applies_to}`stack: ga 9.3` `clivf` for DiskBBQ to store clusters of quantized vectors
* `vex` for the HNSW graph
* `vec` for all non-quantized vector values. This includes all element types: `float`, `byte`, and `bit`.
* `veq` for quantized vectors indexed with [`quantization`](elasticsearch://reference/elasticsearch/mapping-reference/dense-vector.md#dense-vector-quantization): `int4` or `int8`
* `veb` for binary vectors indexed with [`quantization`](elasticsearch://reference/elasticsearch/mapping-reference/dense-vector.md#dense-vector-quantization): `bbq`
* `vem`, `vemf`, `vemq`, and `vemb` for metadata, usually small and not a concern for preloading

> Generally, if you are using a quantized index, you should only preload the relevant quantized values and index structures such as the HNSW graph. Preloading the raw vectors is not necessary and might be counterproductive.

Author: been like that for a bit @shainaraskas but I added some context here; let me know if that reads a bit better now.

Additional detail can be gleened on the specific files by using the [stats endpoint](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-stats) which will display information about the index and fields for example for DiskBBQ you might see something like this:

[source,console]
----
GET my_index/_stats?filter_path=indices.my_index.primaries.dense_vector

> Example Response:

Reviewer: ```suggestion
```

::::

The following file extensions are used for the approximate kNN search: Each extension is broken down by the quantization types.

* {applies_to}`stack: ga 9.3` `cenivf` for DiskBBQ to store centroids
* {applies_to}`stack: ga 9.3` `clivf` for DiskBBQ to store clusters of quantized vectors
* `vex` for the HNSW graph
* `vec` for all non-quantized vector values. This includes all element types: `float`, `byte`, and `bit`.
* `veq` for quantized vectors indexed with [`quantization`](elasticsearch://reference/elasticsearch/mapping-reference/dense-vector.md#dense-vector-quantization): `int4` or `int8`
* `veb` for binary vectors indexed with [`quantization`](elasticsearch://reference/elasticsearch/mapping-reference/dense-vector.md#dense-vector-quantization): `bbq`
* `vem`, `vemf`, `vemq`, and `vemb` for metadata, usually small and not a concern for preloading

Generally, if you are using a quantized index, you should only preload the relevant quantized values and index structures such as the HNSW graph. Preloading the raw vectors is not necessary and might be counterproductive, because paging in the raw vectors might cause the OS to evict important index structures from the cache.

> Additional detail can be gathered about the specific files by using the [stats endpoint](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-stats), which will display information about the index and fields.

Reviewer: ```suggestion
You can gather additional detail about the specific files by using the [stats endpoint](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-stats), which displays information about the index and fields.
```

> For example, for DiskBBQ, you might see something like this:

Reviewer: ```suggestion
For example, for DiskBBQ, the response might look like this:
```
